File: csrc/sequence.cpp
#include "sequence.h"
#include <cassert>
#include <algorithm>
#include <iostream>
// 静态成员初始化
int Sequence::block_size = 256;
int Sequence::counter = 0;



Sequence::Sequence(
    const std::vector<int>& token_ids_,
    py::object sampling_params,
    std::optional<std::any> input_embeds_,
    std::optional<int> seq_id_
)
    : seq_id(seq_id_.value_or(counter++)),
      status(SequenceStatus::WAITING),
      input_embeds(input_embeds_),
      token_ids(token_ids_),
      last_token(token_ids_.empty() ? -1 : token_ids_.back()),
      num_tokens(token_ids_.size()),
      num_prompt_tokens(token_ids_.size()),
      num_cached_tokens(0),
      temperature( sampling_params.attr("temperature").cast<float>() ),   // 从 Python 对象取属性
      max_tokens( sampling_params.attr("max_tokens").cast<int>() ),
      ignore_eos( sampling_params.attr("ignore_eos").cast<bool>() )
{
    block_table.clear();
    local_block_table.clear();
    local_cached_blocks.clear();
    local_uncached_blocks.clear();
    num_blocks_per_slave.clear();
    cum_blocks_per_slave.clear();
    block_belong_to_slave.clear();
    time_usage.clear();
    extra_info.clear();
}

int Sequence::size() const {
    return num_tokens;
}

int Sequence::operator[](int key) const {
    return token_ids.at(key);
}

bool Sequence::is_finished() const {
    return status == SequenceStatus::FINISHED;
}

int Sequence::num_completion_tokens() const {
    return num_tokens - num_prompt_tokens;
}

std::vector<int> Sequence::prompt_token_ids() const {
    if (num_prompt_tokens <= token_ids.size())
        return std::vector<int>(token_ids.begin(), token_ids.begin() + num_prompt_tokens);
    return {};
}

std::vector<int> Sequence::completion_token_ids() const {
    if (num_prompt_tokens <= token_ids.size())
        return std::vector<int>(token_ids.begin() + num_prompt_tokens, token_ids.end());
    return {};
}

int Sequence::num_cached_blocks() const {
    return num_cached_tokens / block_size;
}

int Sequence::num_blocks() const {
    int num_blocks = (num_tokens + block_size - 1) / block_size;
    return num_blocks;
}

int Sequence::length() const {
    return  num_tokens;
}
int Sequence::last_block_num_tokens() const {
    return num_tokens - (num_blocks() - 1) * block_size;
}

std::vector<int> Sequence::block(int i) const {
    assert(i >= 0 && i < num_blocks());
    int start = i * block_size;
    int end = std::min(start + block_size, num_tokens);
    return std::vector<int>(token_ids.begin()+start, token_ids.begin() + end);;
}

void Sequence::append_token(int token_id) {
    token_ids.push_back(token_id);
    last_token = token_id;
    num_tokens++;
}
void Sequence::time_usage_append(double t) {
    time_usage.push_back(t);
}
std::tuple<int, std::optional<std::any>, int, int, int, std::vector<int>, std::any> Sequence::get_state() const {
    if (num_completion_tokens() == 0) {
        return { seq_id, input_embeds, num_tokens, num_prompt_tokens, num_cached_tokens, block_table,
                 std::any(token_ids) };
    } else {
        return { seq_id, input_embeds, num_tokens, num_prompt_tokens, num_cached_tokens, block_table,
                 std::any(last_token) };
    }
}

void Sequence::set_state(const std::tuple<int, std::optional<std::any>, int, int, int, std::vector<int>, std::any>& state) {
    seq_id            = std::get<0>(state);
    input_embeds      = std::get<1>(state);
    num_tokens        = std::get<2>(state);
    num_prompt_tokens = std::get<3>(state);
    num_cached_tokens = std::get<4>(state);
    block_table       = std::get<5>(state);

    if (num_completion_tokens() == 0) {
        token_ids = std::any_cast<std::vector<int>>(std::get<6>(state));
    } else {
        last_token = std::any_cast<int>(std::get<6>(state));
    }
}

File: swiftcache/engine/sequence.py
from copy import copy
from enum import Enum, auto
from itertools import count
from typing import Optional
import time

import torch
from swiftcache.sampling_params import SamplingParams
from swiftcache.utils.io import load_pil_images
from swiftcache.global_config import global_config


class SequenceStatus(Enum):
    WAITING = auto()
    RUNNING = auto()
    FINISHED = auto()

class VisionStatus(Enum):
    WAITING = auto()
    ENCODING = auto()
    COMPLETED = auto()

class RequestType(Enum):
    LANGUAGE_ONLY = auto()
    VISION_LANGUAGE = auto()

class Sequence:
    block_size = global_config.get('kvcache_block_size')
    counter = count()

    def __init__(self, token_ids: list[int], sampling_params = SamplingParams(), input_embeds: Optional[torch.Tensor] = None,seq_id = None):
        self.seq_id = seq_id if seq_id is not None else next(Sequence.counter)
        self.status = SequenceStatus.WAITING
        self.input_embeds = None

        if input_embeds is not None:
            self.input_embeds = input_embeds
        self.token_ids = copy(token_ids)
        self.last_token = self.token_ids[-1]
        self.num_tokens = len(self.token_ids)
        self.num_prompt_tokens = len(self.token_ids)
        self.num_cached_tokens = 0
        self.block_table = []
        self.local_block_table = []
        self.local_cached_blocks = []
        self.local_uncached_blocks = []
        self.num_blocks_per_slave = [] #len(self.num_blocks_per_slave) = len(slave_list) 用来计算local_block_id的偏移量的
        self.cum_blocks_per_slave = [] #len = len(slave_list)+1 
        self.block_belong_to_slave = [] #len = len(self.block_table)
        self.temperature = sampling_params.temperature
        self.max_tokens = sampling_params.max_tokens
        self.ignore_eos = sampling_params.ignore_eos
        self.time_usage = []

        #used for online inference
        self.request_id = None
        self.finised_event = None
        self.extra_info = {}

    def __len__(self):
        return self.num_tokens

    def __getitem__(self, key):
        return self.token_ids[key]

    @property
    def is_finished(self):
        return self.status == SequenceStatus.FINISHED

    @property
    def num_completion_tokens(self):
        return self.num_tokens - self.num_prompt_tokens

    @property
    def prompt_token_ids(self):
        return self.token_ids[:self.num_prompt_tokens]

# 这行有问题吗，明天看一下–
    @property
    def completion_token_ids(self):
        return self.token_ids[self.num_prompt_tokens:]

    @property
    def num_cached_blocks(self):
        return self.num_cached_tokens // self.block_size

    @property
    def num_blocks(self):
        return (self.num_tokens + self.block_size - 1) // self.block_size

    @property
    def last_block_num_tokens(self):
        return self.num_tokens - (self.num_blocks - 1) * self.block_size

    def block(self, i):
        assert 0 <= i < self.num_blocks
        return self.token_ids[i*self.block_size: (i+1)*self.block_size]

    def append_token(self, token_id: int):
        self.token_ids.append(token_id)
        self.last_token = token_id
        self.num_tokens += 1
    
    def time_usage_append(self,t:float):
        self.time_usage.append(t)

    def __getstate__(self):
        return (self.seq_id, self.input_embeds, self.num_tokens, self.num_prompt_tokens, self.num_cached_tokens, self.block_table,
                self.token_ids if self.num_completion_tokens == 0 else self.last_token)

    def __setstate__(self, state):
        self.seq_id, self.input_embeds, self.num_tokens, self.num_prompt_tokens, self.num_cached_tokens, self.block_table = state[:-1]
        if self.num_completion_tokens == 0:
            self.token_ids = state[-1]
        else:
            self.last_token = state[-1]

class VisionSequence:
    counter = count()
    def __init__(self, raw_input=None, sampling_params = SamplingParams(), status:VisionStatus = VisionStatus.WAITING, prepare_inputs = None):
        # assert isinstance(raw_input, dict), f"raw_input must be dict, got {type(raw_input).__name__}"
        self.seq_id = next(Sequence.counter)
        if status is None or status is VisionStatus.WAITING:
            self.status = VisionStatus.WAITING
            self.conversation = raw_input
            self.pil_images = load_pil_images(raw_input)
            self.prepare_inputs: Optional[Any] = None 

        if status is VisionStatus.ENCODING:
            self.status = VisionStatus.ENCODING
            self.prepare_inputs = prepare_inputs

        self.time_usage=[]
        self.sampling_params = sampling_params


def sequence_cpp_to_py(seqs_cpp):
    seqs = []
    for seq_cpp in seqs_cpp:
        sampling_params = SamplingParams(seq_cpp.temperature,seq_cpp.max_tokens,seq_cpp.ignore_eos)
        seq = Sequence(seq_cpp.token_ids,sampling_params,seq_id = seq_cpp.seq_id)
        seq.block_table = seq_cpp.block_table
        seq.num_cached_tokens = seq_cpp.num_cached_tokens
        seqs.append(seq)
    return seqs